core/src/main/java/com/linecorp/armeria/server/Http2StreamLifecycleHandler.java

```diff
@@ -0,0 +1,105 @@
+ *
+ * LY Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+import com.linecorp.armeria.common.HttpRequest;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.util.SafeCloseable;
+import com.linecorp.armeria.internal.common.AbstractHttp2ConnectionHandler;
+
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.http2.Http2ConnectionEncoder;
+import io.netty.handler.codec.http2.Http2Error;
+import io.netty.handler.codec.http2.Http2Stream;
+import io.netty.util.concurrent.EventExecutor;
+import io.netty.util.concurrent.ScheduledFuture;
+
+/**
+ * Synchronizes the lifecycle of an HTTP2 stream with its corresponding user-facing
+ * HTTP constructs. All methods in this class are expected to be invoked from
+ * the {@link Channel}'s {@link EventExecutor}.
+ */
+final class Http2StreamLifecycleHandler implements SafeCloseable {
+
+    private final Http2ConnectionEncoder encoder;
+    private final ChannelHandlerContext ctx;
+
+    private final Map<Integer, ScheduledFuture<?>> streamResetFutures = new HashMap<>();
+
+    Http2StreamLifecycleHandler(ChannelHandlerContext ctx,
+                                AbstractHttp2ConnectionHandler handler) {
+        encoder = handler.encoder();
+        this.ctx = ctx;
+    }
+
+    /**
+     * Invoked when a {@link Http2Stream}'s corresponding {@link HttpRequest} and {@link HttpResponse}
+     * are closed.
+     */
+    void maybeResetStream(int streamId, Http2Error http2Error, long delayMillis) {
+        if (delayMillis == 0) {
+            maybeResetStream0(streamId, http2Error);
+        } else if (delayMillis > 0) {
+            final ScheduledFuture<?> scheduled = ctx.executor().schedule(() -> {
+                maybeResetStream0(streamId, Http2Error.CANCEL);
+            }, delayMillis, TimeUnit.MILLISECONDS);
+            streamResetFutures.put(streamId, scheduled);
+        }
+    }
+
+    private void maybeResetStream0(int streamId, Http2Error http2Error) {
+        if (!ctx.channel().isActive()) {
+            return;
+        }
+        final Http2Stream stream = encoder.connection().stream(streamId);
+        if (stream == null) {
+            return;
+        }
+        if (stream.state().remoteSideOpen()) {
```

Comment:
Should we check these conditions before scheduling `maybeResetStream0()` as well?